### ISSUE TITLE
Minification corrupts return path

### ISSUE THREAD

**Issue by lramos15** [evanw/esbuild#1350]
VS Code uses esbuild for minification and it normally works flawlessly. However we have a simple undefined check that if undefined should return and the minification appears to just skip over that return. I'm not really sure why and any insight would be helpful. Instead of reiterating the issue I'll just link the VS Code issue here https://github.com/microsoft/vscode/issues/125518#issuecomment-854790533

**Comment by evanw** [comment]
So the code in question is this:

```js
if (editor instanceof DiffEditorInput) {
  if (!selectedContribution.createDiffEditorInput) {
    return;
  }
  const inputWithOptions = selectedContribution.createDiffEditorInput(editor, options, group);
  return inputWithOptions;
}
```

With `--minify-syntax` that becomes this:

```js
if (editor instanceof DiffEditorInput)
  return selectedContribution.createDiffEditorInput ? selectedContribution.createDiffEditorInput(editor, options, group) : void 0;
```

This seems equivalent to me. Here are the transformations that lead to that result:

1. The original code
    ```js
    if (editor instanceof DiffEditorInput) {
      if (!selectedContribution.createDiffEditorInput) {
        return;
      }
      const inputWithOptions = selectedContribution.createDiffEditorInput(editor, options, group);
      return inputWithOptions;
    }
    ```

2. Inline `inputWithOptions`
    ```js
    if (editor instanceof DiffEditorInput) {
      if (!selectedContribution.createDiffEditorInput) {
        return;
      }
      return selectedContribution.createDiffEditorInput(editor, options, group);
    }
    ```

3. Combine into `if`/`else`
    ```js
    if (editor instanceof DiffEditorInput) {
      if (!selectedContribution.createDiffEditorInput) {
        return;
      } else {
        return selectedContribution.createDiffEditorInput(editor, options, group);
      }
    }
    ```

4. Turn the `if` statement into a conditional expression
    ```js
    if (editor instanceof DiffEditorInput) {
      return !selectedContribution.createDiffEditorInput ? void 0 : selectedContribution.createDiffEditorInput(editor, options, group);
    }
    ```

5. Remove the `!` operator by swapping the true and false branches
    ```js
    if (editor instanceof DiffEditorInput) {
      return selectedContribution.createDiffEditorInput ? selectedContribution.createDiffEditorInput(editor, options, group) : void 0;
    }
    ```

6. Unwrap the block statement
    ```js
    if (editor instanceof DiffEditorInput)
      return selectedContribution.createDiffEditorInput ? selectedContribution.createDiffEditorInput(editor, options, group) : void 0;
    ```

What do you think? Does that sound right?

Also the image linked in https://github.com/microsoft/vscode/issues/125518#issuecomment-854790533 appears to show evaluation inside the return statement, but then it continues afterward. That seems like the core of the problem to me; regardless of how that return statement got there (i.e. even if there was a minification bug in esbuild), a return statement is not returning.

Some thoughts that come to mind:

* Maybe developer tools has a bug with the mapping of automatically-formatted code back to the actually-running code? Or some other bug with breakpoints that is making things look different than they are?
* Maybe something with source maps is messing something up? It doesn't seem so since this is minified code, but you never know.

Some things to try:

* Does this reproduce with minification off? If so, this isn't a problem with minification.
* Instead of passing `--minify` you could try passing `--minify-syntax`, `--minify-identifiers`, and `--minify-whitespace` individually to narrow down the problem further (see [the docs](https://esbuild.github.io/api/#minify) for more info).

**Comment by bpasero** [comment]
👋  @evanw apologies for this false assumption, I had originally thought it was a minification issue but it turns out not to be an issue with ESBuild. 

The fact that devtools was jumping over the `return` statement made be believe that something had to be wrong with the minified code but then I was able to reproduce without minification (only today because I was in a hurry yesterday).

We are currently building VSCode with ESBuild `0.12.1` and will ship our stable release next week with it. In our previous release we shipped with ESBuild `0.8.30` and since this is such a big jump in versions I was very nervous that we could see issues related to minification. 

Since we are wrapping up our stable release, we will probably stay on `0.12.1` and only update to `0.12.6` the month after unless of course you think the fixes that went into ESBuild meanwhile are critical. We have a bit of confidence with `0.12.1` due to our ~30k insider users of VSCode that have already been running with the updated ESBuild for >1 week.

Keep up the great work ❤️

**Comment by evanw** [comment]
Ah, I see. In that case I'm going to close this issue since it sounds like it's working as intended.

The fixes that have gone in since 0.12.1 are for some pretty obscure edge cases so you are probably fine sticking with 0.12.1. The only change you might potentially want to pick up is a non-determinism fix that landed in 0.12.4. Large code bases that used `import()` expressions and had esbuild's code splitting mode active could result in repeated builds returning output that behaved the same but that had different hashes in the file names for different runs. You would only care about that if you are using esbuild's code splitting mode though, and only if consistent hashes in file names is important to you.

**Comment by bpasero** [comment]
> You would only care about that if you are using esbuild's code splitting mode though, and only if consistent hashes in file names is important to you.

I see. I think we are [not using](https://github.com/microsoft/vscode/blob/9cba86888c953fe4f665290462b2bd701f3cb971/build/lib/optimize.ts#L252-L258) the code splitting feature for VSCode.
